### ISSUE TITLE
Error details: model expected the shape of dimension 0 to be between 1 and 1 but received 5

### ISSUE THREAD

**Issue by royinx** [triton-inference-server/server#3032]
**Description:**

Raise `Error details: model expected the shape of dimension 0 to be between 1 and 1 but received 5`
when sending a request in batch of 5 images, 

Server: [details](https://github.com/royinx/trtis_issues2/blob/main/server/models/model_x/config.pbtxt)
dynamic batching model with
```
name: "model_x"
platform: "tensorrt_plan"
max_batch_size: 64

dynamic_batching {
  preferred_batch_size: [ 8, 16, 32, 64 ]
  max_queue_delay_microseconds: 100
}

```

**Triton Information**
`<21.02-py3>` & `<21.02-py3-sdk>` 

**Reproduce issues**
```bash
git clone https://github.com/royinx/trtis_issues2
```
```bash
# go to model folder 
# using nvcr.io/nvidia/tensorrt:21.02-py3 to compile
docker run  --rm \
            -it \
            --runtime nvidia \
            -v ${PWD}:/sharefolder \
            -w /sharefolder \
            nvcr.io/nvidia/tensorrt:21.02-py3 \
trtexec --onnx=model.onnx --saveEngine=model.plan --explicitBatch --fp16 --verbose --dumpOutput --minShapes='input':1x3x480x640 --optShapes='input':5x3x480x640  --maxShapes='input':64x3x480x640
```
```
docker-compose build && docker-compose up

# go into the client controller
docker exec -it controller bash
# batch size = 1 , Work
python3 test.py 1
# batch size = 5 , Raised
python3 test.py 5 
```

---

### Another Questions

1. I have to change [`input_layer["dims"]`](https://github.com/royinx/trtis_issues2/blob/27065de35a3aa7e2ec7257c20d32b58ea9b6d2b6/client/test.py#L21) to fulfil the model expectations every request, any better solution?
In 19.12, batch_size can be specified in [`ctx.run()`](https://github.com/triton-inference-server/server/blob/a1f3860ba65c0fd8f2be3adfcab2673efd039348/src/clients/python/simple_client.py#L89) function.
since 21.02 doesn't have batch_size, [`httpclient.InferInput.set_data_from_numpy()`](https://github.com/triton-inference-server/client/blob/23de3ad01e9bc22dbbf8a30d8ebfbffc51110b31/src/python/library/tritonclient/http/__init__.py#L1567) will check for the shape expectation from TritonIS. 
Keep checking batch_size on every request is inconvenient for dynamic batching.

2.  May I request an example on sending a batch of images to a dynamic batching server?
including `Trtexec compile command`, `config.pbtxt` and `client.infer function`

**Comment by Tabrizian** [comment]
Thanks for providing the details.

1. When dynamic batching is enabled, you need to add another dimension indicating the size of your batch. If the batch size is 1, the first dimension of your input should contain a single element. The error that your seeing is because of this. You can find more information about it [here](https://github.com/triton-inference-server/server/blob/main/docs/model_configuration.md#inputs-and-outputs).
2. You can find more information about the [trtexec](https://github.com/NVIDIA/TensorRT/blob/master/samples/opensource/trtexec/README.md#using-trtexec), [Triton Model Config](https://github.com/triton-inference-server/server/blob/r21.05/docs/model_configuration.md), and [examples for Python client](https://github.com/triton-inference-server/client/tree/r21.05/src/python/examples).

**Comment by deadeyegoodwin** [comment]
Closing due to inactivity.

**Comment by royinx** [comment]
According to `When dynamic batching is enabled, you need to add another dimension indicating the size of your batch.`
you mean i hv to add the dimension in `infer()` or `config.pbtxt` ?

**Comment by GuanLuo** [comment]
If your model config is similar to
```
max_batch_size: 64
input [ { name: "input" dims: [2, 3, 4, 5] data_type: TYPE_FP32}, ...]
...
```
And you are sending a batch 8 request, then you should invoke the InferInput constructor as the following:
```
grpcclient.InferInput('input', [8, 2, 3, 4, 5], "FP32")
```

**Comment by royinx** [comment]
Thanks @GuanLuo and @Tabrizian. I figured out the problem. 
Both .pbtxt , InferInput() and infer() is configuring correctly. 

Here is my debug process:
run `tritonserver --model-repository=/models --strict-model-config=false` 
give  `unable to autofill for 'model', configuration specified max-batch 64 but TensorRT engine only supports max-batch 1`

So I think it is the root cause of error. 
I tested with 
`perf_client -m <model> -u trtis:8000 -i HTTP -v -p3000 -d -l3000 -t1 -c5 -b5 `  and return same error ` between 1 and 1 but received 5`, means error is not from TRTIS

So the problem is from TensorRT .plan file.
According to [TensorRT #1111](https://github.com/NVIDIA/TensorRT/issues/1111) and [Triton #2606](https://github.com/triton-inference-server/server/issues/2606), I found the engine is built from `N=1` .onnx file 
Change to dynamic variable in onnx , then rebuild the engine with 
```
trtexec --onnx=model.onnx \
            --explicitBatch \
            --fp16 --verbose \
            --minShapes="data":1x3x480x640 \
            --optShapes="data":5x3x480x640 \
            --maxShapes="data":64x3x480x640 \
            --shapes="data":5x3x480x640 \
            --saveEngine=model.plan
```

---


During testing the IS , I found out a question:
- I use the engine built in implicit mode, max batch size = 1, <br/>  while .pbtxt set `max_batch_size:64` + `dynamic:[8,16,32,64]`  <br/> IS still can load the engine well and show `READY` , also can do inference.  <br/> Just the client can only send 1 img / batch .
